internal/freelist/serde.go

```diff
@@ -0,0 +1,79 @@
+package freelist
+
+import (
+	"fmt"
+	"sort"
+	"unsafe"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type Serializable interface {
+	// Read calls Init with the page ids stored in te given page.
+	Read(f Interface, page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(f Interface, page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize(f Interface) int
+}
```

Comment:
Propose to change the interface something like below,
```
type ReadWriter interface {
	// Read initializes the freelist from the given page.
	Read(page *common.Page)

	// Write writes the freelist into the given page.
	Write(page *common.Page)

	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
	// This should never underestimate the size.
	EstimatedWritePageSize() int
}
```

Also embedded it into Interface,
```
type Interface interface {
	ReadWriter
	...... 
}
```

Reply:
+1 on the naming, the pull into the interface has some coupling issues though. When we move it into `shared` (which would make sense), the following functions are missing from the shared struct: `Count`, `Init` and `CopyAll` would also need to become part of the implementation in shared (which again depends on `PendingCount` and `FreePageIds`).

I think ultimately `shared` will just become the old freelist.go if we continue down that path ;-)

Otherwise we can move the three read/write methods copied into both array+hashmap, which also resolves it.

WDYT?

Reply:
also, maybe a more general question: should the serialization be implementation dependent or interface-dependent? 

Currently we have to keep the serde the exact same between array/hashmap, otherwise we won't be able to switch the freelist implementation "at runtime" after a bbolt has been created. If we make it implementation-dependent, we need to devise a way to transform one to the other.

Reply:
I've mangled it into the shared interface, similar to mergeSpans, added it into a separated commit for better comparison.

Reply:
I've saved the commit into another branch https://github.com/tjungblu/bbolt/tree/refactor_freelists_commits

squashed this one here to have an easier time rebasing on top of #783